### PR TITLE
Feature/proj 120 token handling

### PR DESCRIPTION
**Pull request by sven-hm**

### CODE REVIEW

File: register_thing.c [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+static const char *TAG = "UBIRCH REGISTER THING";
+
+/*
+curl --location --request POST \
+   'https://api.console.prod.ubirch.com/ubirch-web-ui/api/v1/devices/create?with_api_info=true' \
+   --header 'Content-Type: application/json' \
+   --header 'Authorization: Bearer <TOKEN>' \
+   --data-raw '{"hwDeviceId":"<UUID>", "description":"ADD DESCRIPTION OF DEVICE HERE"}'
+*/
```
Comment (sven-hm) [inline register_thing.c:17]: i'll remove this

File: test/test_register_thing.c [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    const size_t data_len = UBIRCH_REGISTER_THING_JSON_OBJECT_SIZE + strlen(description);
+    char* string_buffer = malloc(data_len);
+    // note: build_post_json_object returns number of characters written without terminatin zero
+    TEST_ASSERT_EQUAL_INT(data_len - 1, build_post_json_object(uuid, description, string_buffer, data_len));
+    TEST_ASSERT_EQUAL_STRING(
+            "{\"hwDeviceId\":\"00010203-0405-0607-0809-0a0b0c0d0e0f\",\"description\":\"hello world\"}",
+            string_buffer);
+}
+
+TEST_CASE("parse api info", "[register_thing]") {
```
Comment (gruenwaldi) [inline test/test_register_thing.c:21]: parsing failure not tested, but lets keep it for now. Maybe add TODO

**Review by gruenwaldi — CHANGES_REQUESTED** [review]
Looks good in General.

Please check the comments and resolve

File: copyrt.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/*
```
[the remark this answers is not in the document]
Comment (sven-hm) [reply copyrt.h:1]: I'm not sure. I copied it over here as it was included from the api-helper header.

File: register_thing.c [hunk 4]
```diff
@@ -0,0 +1,203 @@
+#define LOG_LOCAL_LEVEL ESP_LOG_DEBUG
+
+#include <string.h>
+#include <register_thing.h>
+#include <esp_http_client.h>
+#include <esp_log.h>
+#include "api-http-helper.h"
+#include <cJSON.h>
+#include <id_handling.h>
+#include <token_handling.h>
```
[the remark this answers is not in the document]
Comment (sven-hm) [reply register_thing.c:10]: fixed

File: register_thing.c [hunk 5]
```diff
@@ -0,0 +1,203 @@
+#define LOG_LOCAL_LEVEL ESP_LOG_DEBUG
```
[the remark this answers is not in the document]
Comment (sven-hm) [reply register_thing.c:1]: added

File: register_thing.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+ * Parse json and check if it matches configuration. If everything is fine,
+ * write password into password_buffer.
+ */
+static int parse_api_info(const unsigned char* expected_uuid, const char* json,
+        size_t json_size, char* password_buffer, size_t password_buffer_len) {
+    ESP_LOGI(TAG, "parsing api info");
+    ESP_LOG_BUFFER_HEXDUMP(TAG, json, (uint16_t) json_size, ESP_LOG_DEBUG);
+    char expected_uuid_string[37];
+    if (uuid_to_string(expected_uuid, expected_uuid_string, 37) < 0) {
+        return -1;
```
[the remark this answers is not in the document]
Comment (sven-hm) [reply register_thing.c:52]: done

File: register_thing.c [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        ESP_LOGE(TAG, "failed to read password");
+        goto PARSE_API_INFO_ERROR;
+    }
+    if (snprintf(password_buffer, password_buffer_len, password->valuestring) < 0) {
+        ESP_LOGE(TAG, "password buffer too small");
+        goto PARSE_API_INFO_ERROR;
+    }
+
+    cJSON_Delete(api_info);
+    return 0;
```
[the remark this answers is not in the document]
Comment (sven-hm) [reply register_thing.c:111]: done

File: register_thing.c [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+#include <cJSON.h>
+#include <id_handling.h>
+#include <token_handling.h>
+
+static const char *TAG = "UBIRCH REGISTER THING";
+
+/*
+ * Parse json and check if it matches configuration. If everything is fine,
+ * write password into password_buffer.
+ */
```
[the remark this answers is not in the document]
Comment (sven-hm) [reply register_thing.c:45]: done

File: test/test_register_thing.c (hunk 2 above)
Comment (sven-hm) [reply test/test_register_thing.c:21]: Added some more tests. While testing i found another issue with `snprintf`, this is fixed as well.

**Review by gruenwaldi — APPROVED** [review]

**Pull request merged**